azuredevops/internal/service/approvalsandchecks/resource_check_required_template.go

```diff
@@ -0,0 +1,104 @@
+package approvalsandchecks
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/utils/pipelineschecksextras"
+)
+
+var validRepositoryTypes = []string{"git", "github", "bitbucket"}
+
+// ResourceCheckRequiredTemplate schema and implementation for required template check resources
+func ResourceCheckRequiredTemplate() *schema.Resource {
+	r := genBaseCheckResource(flattenCheckRequiredTemplate, expandCheckRequiredTemplate)
+
+	r.Schema["required_template"] = &schema.Schema{
```

Comment:
The idea of the nested block was to allow users to specify more than one required template, following the api contract.

I put an example in the docs:
```hcl
resource "azuredevops_check_required_template" "example" {
  project_id           = azuredevops_project.example.id
  target_resource_id   = azuredevops_environment.example.id
  target_resource_type = "environment"

  required_template {
    repository_name = "project/repository"
    repository_ref  = "refs/heads/main"
    template_path   = "template/path.yml"
  }

  required_template {
    repository_name = "project/repository"
    repository_ref  = "refs/heads/main"
    template_path   = "template/alternate-path.yml"
  }
}
```

I guess for most use cases only one required template will be specified but since the api allowed for more than one I went with it.

But if you think it's better design to keep everything in the root schema and allow for only 1 required template to be set please let me know. :)

Reply:
1. Add  a `MinItems: 1,` constraint to ensure at lease one  template configured. 
2. Since all the properties within `required_template` can be updated, suggest  change the type from `TypeSet` to `TypeList`  if service will not change the request template order. Because type set will generate the index based on the hash code, as a result any change of the `required_template` block will like:
  ![image](https://github.com/microsoft/terraform-provider-azuredevops/assets/57888764/110ac51f-0b81-48ee-a471-be4dc8cccf79)

Reply:
Done.

One scenario that might be confusing is the following:
1. create resource with 2 `required_template` blocks: 
```hcl
resource "azuredevops_check_required_template" "example_vg" {
  project_id           = azuredevops_project.example.id
  target_resource_id   = azuredevops_variable_group.example.id
  target_resource_type = "variablegroup"
  required_template {
    repository_type = "azuregit"
    repository_name = "templates"
    repository_ref  = "refs/heads/master"
    template_path   = "templates/azure-pipelines.yml"
  }
  required_template {
    repository_type = "azuregit"
    repository_name = "templates"
    repository_ref  = "refs/heads/master"
    template_path   = "templates/alternate-azure-pipelines.yml"
  }
}
```
2. delete the first one from the list:
```hcl
resource "azuredevops_check_required_template" "example_vg" {
  project_id           = azuredevops_project.example.id
  target_resource_id   = azuredevops_variable_group.example.id
  target_resource_type = "variablegroup"
  required_template {
    repository_type = "azuregit"
    repository_name = "templates"
    repository_ref  = "refs/heads/master"
    template_path   = "templates/alternate-azure-pipelines.yml"
  }
}
```
as a result the plan will be like:
![image](https://github.com/microsoft/terraform-provider-azuredevops/assets/23347635/7cec656c-0b85-46d1-bfe6-ba32890dd1d3)

as if you were deleting the second one and changing the first one. But I guess that's expected.

Reply:
Usually this kind of constraint  relies strongly on the configuration order,   update the template without changing the order is acceptable.